vignettes/cpp11_faq.Rmd

```diff
@@ -0,0 +1,236 @@
+---
+title: "cpp11 FAQ"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{cpp11 FAQ}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+library(cpp11)
+```
+
+Intended to answer common questions about the usage of cpp11 objects
+
+## FAQ
+
+#### 1. How do I add elements to a named list?
+
+  * Use push_back() with named literals to add elements to a named list
+  
+```{cpp11}
+#include <cpp11.hpp>
+
+using namespace cpp11;
+
+
+[[cpp11::register]]
+cpp11::list foo_push() {
+  cpp11::writable::list x;
+  x.push_back({"foo"_nm = 1});
+  return x;
+}
+```
+
+#### 2. How do I create a new empty environment?
+
+  * Use the C++ `auto` keyword with R's `new.env` to return a new empty environment. This should then be passed to `cpp11::environment` before use
+
+```{cpp11}
+#include <cpp11.hpp>
+
+using namespace cpp11;
+
+[[cpp11::register]]
+cpp11::environment get_environment() {
+  auto new_env = cpp11::package("base")["new.env"];
+  cpp11::environment my_env(new_env());
+  return my_env;
+}
+
+```
+
+#### 3. How do I assign and retrieve values in an environment? What happens if I try to get a value that doesn't exist?
+
+  * To assign and retrieve values in an environment use brackets to access values in your environment by name.
+  * To check if a named value exists use `my_env["bar"] == R_UnboundValue` which returns `TRUE` if the value is unassigned and NULL.
+
+```{cpp11}
+#include <cpp11.hpp>
+
+using namespace cpp11;
+[[cpp11::register]]
+cpp11::r_bool foo_exists(){
+  auto new_env = cpp11::package("base")["new.env"];
+  cpp11::environment my_env(new_env());
+  
+  my_env["foo"] = 5;
+  cpp11::r_bool fofo = (as_cpp<int>(my_env["foo"]) == 5);
+  
+  return (fofo);
+}
+```
+
+```{cpp11}
+#include <cpp11.hpp>
+
+using namespace cpp11;
+
+[[cpp11::register]]
+cpp11::r_bool bar_exists(){
+  auto new_env = cpp11::package("base")["new.env"];
+  cpp11::environment my_env(new_env());
+  
+  
+  cpp11::r_bool barbar = (my_env["bar"] == R_UnboundValue);
+  
+  return (!barbar);
+}
+```
+
+#### 4. How can I create a `cpp11:raws` from a `std::string`?
+
+  * Push back the contents individually
+  
+```{cpp11}
+#include <cpp11.hpp>
+
+using namespace cpp11;
+[[cpp11::register]]
+cpp11::raws push() {
+  
+    std::string x("hi");
+    cpp11::writable::raws out;
+  
+    for (auto c : x) {
+      out.push_back(c);
+    }
+  
+    return out;
+  }
+```
+  
+#### 5. What are the underlying types of cpp11 objects?
+
+   | vector   | element
+   | ---      | ---
+   | integers |	int
+   | doubles	| double
+   | logical  | r_bool
+   | strings	| r_string
+   | raws	    | uint8_t
+   |  list	  | SEXP
+
+#### 6. How do I create a new empty list?
+
+`cpp11::writable::list x;`
+
+#### 7. How do I retrieve (named) elements from a named vector/list?
+
+`x["foo"]`
+
+#### 8. How can I tell whether a vector is named?
+
+  TODO    
```

Comment:
We can update this now with `x.named()` right?

Reply:
Yep! 👍